File: 2015/day12.js
const data = require('./input/day12')

const string = JSON.stringify(data)

const list = string.replace(/[^0-9,-]/g,'').replace(/,{2,}/g,',')

const array = JSON.parse(`[${list}]`).map(Number)
console.log(array.reduce((total, n) => total + n))

const sum = object => {
  if (typeof object === 'string') {
    return 0
  } else if (typeof object === 'number') {
    return object
  } else if (Array.isArray(object)) {
    return object.reduce((total, n) => total + sum(n), 0)
  } else if (Object.values(object).includes('red')) {
    return 0
  } else {
    return Object.keys(object).reduce((total, key) => total + sum(object[key]), 0)
  }
}
console.log(sum(data))

File: 2015/day12.rb
require 'json'
data = JSON.parse(File.read('./input/day12.json'))

def sum(obj)
  if obj.is_a? Numeric
    obj
  elsif obj.is_a?(String) || obj.nil?
    0
  elsif obj.is_a? Array
    obj.reduce(0){|total, elm| total + sum(elm) }
  elsif obj.values.any?{|key| key == 'red'}
    0
  else
    obj.values.reduce(0){|total, elm| total + sum(elm) }
  end
end

puts sum(data)
